src/sign/records.rs

```diff
@@ -243,6 +256,239 @@ impl<N, D> SortedRecords<N, D> {
+            // are done.
+            if !family.is_in_zone(apex) {
+                break;
+            }
+
+            // If the family is below a zone cut, we must ignore it.
+            if let Some(ref cut) = cut {
+                if family.owner().ends_with(cut.owner()) {
+                    continue;
+                }
+            }
+
+            // A copy of the family name. We’ll need it later.
+            let name = family.family_name().cloned();
+
+            // If this family is the parent side of a zone cut, we keep the
+            // family name for later. This also means below that if
+            // `cut.is_some()` we are at the parent side of a zone.
+            cut = if family.is_zone_cut(apex) {
+                Some(name.clone())
+            } else {
+                None
+            };
+
+            // RFC 5155 7.1 step 2:
+            //   "If Opt-Out is being used, owner names of unsigned
+            //    delegations MAY be excluded."
+            let has_ds = family.records().any(|rec| rec.rtype() == Rtype::DS);
+            if cut.is_some() && !has_ds && opt_out {
+                continue;
+            }
+
+            // RFC 5155 7.1 step 4:
+            //   "If the difference in number of labels between the apex and
+            //    the original owner name is greater than 1, additional NSEC3
+            //    RRs need to be added for every empty non-terminal between
+            //    the apex and the original owner name."
+            let distance_to_root = name.owner().iter_labels().count();
+            let distance_to_apex = distance_to_root - apex_label_count;
+            if distance_to_apex > 1 {
+                // Are there any empty nodes between this node and the apex?
+                // The zone file records are already sorted so if all of the
+                // parent labels had records at them, i.e. they were non-empty
+                // then non_empty_label_count would be equal to label_distance.
+                // If it is less that means there are ENTs between us and the
+                // last non-empty label in our ancestor path to the apex.
+
+                // Walk from the owner name down the tree of labels from the
+                // last known non-empty non-terminal label, extending the name
+                // each time by one label until we get to the current name.
+
+                // Given a.b.c.mail.example.com where:
+                //   - example.com is the apex owner
+                //   - mail.example.com was the last non-empty non-terminal
+                // This loop will construct the names:
+                //   - c.mail.example.com
+                //   - b.c.mail.example.com
+                // It will NOT construct the last name as that will be dealt
+                // with in the next outer loop iteration.
+                //   - a.b.c.mail.example.com
+                for n in (1..distance_to_apex - 1).rev() {
```

Comment:
Thanks! I was planning to add proper test cases but for now have extended my test input zone to cover this missing case and realize now why I had some additional logic in there before which I removed because I couldn't see what value it was adding... 😛